src/CodeViewer/include/CodeViewer/Dialog.h

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef CODE_VIEWER_DIALOG_H_
+#define CODE_VIEWER_DIALOG_H_
+
+#include <CodeViewer/Em.h>
+#include <CodeViewer/Viewer.h>
+
+#include <QDialog>
+#include <QSyntaxHighlighter>
+#include <memory>
+#include <optional>
+
+namespace Ui {
+class CodeViewerDialog;  // IWYU pragma: keep
+}
+
+namespace orbit_code_viewer {
+
+/*
+  This is a dialog for displaying source code (and assembly to be strict).
+
+  The typical use-case is to instantiate locally and call the exec function
+  which blocks until the dialog is closed:
+
+  orbit_code_viewer::Dialog dialog{};
+  dialog.SetSourceCode(source_code);
+  dialog.exec();
+
+  Optionally a syntax highlighter can be provided with the source code.
+  Check out the Syntax highlighting module for more details on this.
+  Example:
+
+  orbit_code_viewer::Dialog dialog{};
+  dialog.SetSourceCode(source_code, std::make_unique<orbit_syntax_highlighter::X86Assembly>());
+  dialog.exec();
+
+  Check out orbit_code_viewer::Viewer if you don't need a dialog but rather a widget
+  which can be embedded into other windows.
+*/
+class Dialog : public QDialog {
+  Q_OBJECT
+
+ public:
+  explicit Dialog(QWidget* parent = nullptr);
+  ~Dialog() noexcept override;
+
+  void SetSourceCode(const QString& code);
+  void SetSourceCode(const QString& code, std::unique_ptr<QSyntaxHighlighter> syntax_highlighter);
```

Comment:
We have a lot of problems in the code base due to nullptr. One reason for that is that you never know if the function that you're calling accepts a nullptr or not.

My rule of thumb: std::unique_ptr or a raw pointer on interface boundaries can NEVER be nullptr! This makes it a lot simpler to reason about the code.

If I want to also accept nullptr I either wrap it into a `std::optional` or add overloads. And since `std::optional<std::unique_ptr<...>>` is pretty ugly, I prefer overloads in this case.

Reply:
Perfect Henning, thanks for explaining me :)